vtysh/vtysh.c

```diff
@@ -725,19 +725,17 @@ int vtysh_mark_file(const char *filename)
 		switch (vty->node) {
 		case LDP_IPV4_IFACE_NODE:
 			if (strncmp(vty_buf_copy, "   ", 3)) {
-				vty_out(vty, "  end\n");
```

Comment:
This part makes me nervous, we have a cubic ton of bugs in the past around these vtysh hacks. What is this doing exactly?

Reply:
A typical ldpd config might look something like this:
```
mpls ldp
 router-id 10.0.0.1
 !
 address-family ipv4
  discovery transport-address 10.0.0.1
  !
  interface eth0
  !
  interface eth1
  !
  interface eth2
  !
 exit-address-family
 !
!
```
This enables ldp for the 3 interfaces `eth0`, `eth1` and `eth2`. When calling `vtysh -m` on such a config file, the line of code deleted in the commit above would add an `end` clause after each of the `interface ...` lines. This will make the reload script erroneously believe that the `mpls ldp` context ends at the first interface, which of course messes up the diff computation.

Reply:
I see. Can we safely change this from `end` to `exit` rather than removing it, right? This way we still maintain the walkup behavior but end up in the correct node.

Reply:
That will not work either, as the script will believe that the `exit` marks the end of the `address-family ipv4` subcontext. For it to work, we would need to also enhance the script with additional special rules to understand that `interface eth0`, in the context of `address family ipvX` inside `mpls ldp` is the start of a new (and most of the times empty) sub-subcontext. 

See this example output from the script when adding those `exit` commands back:
```
2019-11-12 14:05:21,491 DEBUG: LINE mpls ldp                                          : entering new context, [u'mpls ldp']
2019-11-12 14:05:21,491 DEBUG: LINE router-id 7.0.0.1                                 : append to current_context_lines, [u'mpls ldp']
2019-11-12 14:05:21,491 DEBUG: LINE address-family ipv4                               : entering sub-context, append to ctx_keys
2019-11-12 14:05:21,491 DEBUG: LINE discovery transport-address 7.0.0.1               : append to current_context_lines, [u'mpls ldp', u'address-family ipv4']
2019-11-12 14:05:21,491 DEBUG: LINE interface 1-xge2.0                                : append to current_context_lines, [u'mpls ldp', u'address-family ipv4']
2019-11-12 14:05:21,491 DEBUG: LINE exit                                              : popping from subcontext to ctx[u'mpls ldp']
2019-11-12 14:05:21,491 DEBUG: LINE interface 1-xge3.0                                : append to current_context_lines, [u'mpls ldp']
2019-11-12 14:05:21,491 DEBUG: LINE exit                                              : popping from subcontext to ctx[u'mpls ldp']
2019-11-12 14:05:21,491 DEBUG: LINE interface 1-xge4.0                                : append to current_context_lines, [u'mpls ldp']
2019-11-12 14:05:21,491 DEBUG: LINE exit                                              : popping from subcontext to ctx[u'mpls ldp']
2019-11-12 14:05:21,492 DEBUG: LINE exit-address-family                               : popping from subcontext to ctx[u'mpls ldp']
2019-11-12 14:05:21,492 DEBUG: LINE end                                               : exiting old context, [u'mpls ldp']
```

Reply:
@qlyoung  I just pushed a new commit that does exactly that, in case you prefer this flavor. It is probably a better solution than the previous one, even though it adds complexity to the script.